### PR TITLE
New core typechecker

### PR DESCRIPTION
**Pull request by andy-morris**

### CODE REVIEW

**Comment by andy-morris** [comment]
**it is done** 🎉🎉🎉

this is a huge PR, sorry about that. the main parts of interest are in the typechecker and erasure, of course; the rest is mostly fixing up call sites since the typechecker now returns leftovers where necessary, and the erasure now returns just a single annotated term.

File: app/Interactive.hs [hunk 1]
```diff
@@ -73,22 +72,23 @@ handleSpecial str cont =
...
-                  ExecTerm Nat.Ty Nat.Val ()
-              )
-          H.outputStrLn (show erased)
-          case erased of
-            (Right (Core.Assignment term _), _) ->
-              transformAndEvaluateErasedCore Nat.t True term
-            _ -> return ()
-        _ -> H.outputStrLn "must enter a valid annotated core term"
+    'c' : 't' : ' ' : _rest -> do
+      H.outputStrLn "TODO fix typecheckAffineErase"
```
Comment (cwgoes) [inline app/Interactive.hs:76]: ... yeah this isn't too important for now

**Comment by cwgoes** [comment]
@andy-morris If you resolve the conflicts I can merge this, since I want to work off this for the datatypes stuff.

**Comment by andy-morris** [comment]
done! I ran the tests locally just now to make sure I (or git) didn't mess it up, but we can wait for the ci too if you want.

**Pull request merged**